src/Build.UnitTests/Graph/IsolateProjects_Tests.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using Microsoft.Build.UnitTests;
+using Shouldly;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Build.Graph.UnitTests
+{
+    public class IsolateProjectsTests
+    {
+        private readonly string _project = @"
+                <Project DefaultTargets='BuildSelf'>
+                    <Target Name='BuildReference'>
+                        <MSBuild Projects='{0}' Targets='ReferenceTarget'/>
+                    </Target>
+
+                    <Target Name='BuildSelf'>
+                        <MSBuild Projects='$(MSBuildThisFile)' Targets='SelfTarget'/>
+                    </Target>
+
+                    <Target Name='SelfTarget'>
+                    </Target>
+                </Project>";
+
+        private readonly string _reference = @"
+                <Project>
+                    <Target Name='ReferenceTarget'>
+                    </Target>
+                </Project>";
+
+        private readonly ITestOutputHelper _testOutput;
+
+        public IsolateProjectsTests(ITestOutputHelper testOutput)
+        {
+            _testOutput = testOutput;
+        }
+
+        [Theory]
+        [InlineData(BuildResultCode.Success, new string[] {})]
+        [InlineData(BuildResultCode.Success, new[] {"BuildSelf"})]
+        public void CacheEnforcementShouldAcceptSelfReferences(BuildResultCode expectedBuildResult, string[] targets)
+        {
+            AssertBuild(targets,
+                (result, logger) =>
+                {
+                    result.OverallResult.ShouldBe(BuildResultCode.Success);
+
+                    logger.ErrorCount.ShouldBe(0);
```

Comment:
I wonder if `logger.Errors.ShouldBeEmpty()` would be more diagnosable when it fails. As is it'd say "1 error, expected 0", and the natural question is "so what was the error?"